Ensure that the rock image can be used as drop-in replacement

petrutlucian94 opened the pull request:
rawfile-localpv [launches pods on the fly when performing resizes](https://github.com/canonical/rawfile-localpv/blob/main/templates/task.yaml) . This operation currently fails due to the following:

* the image repo and tag env variables aren't specified
  * [the docker file uses build arguments to set these](https://github.com/canonical/rawfile-localpv/blob/3bcdc388a39d9352c1cbe652965c268bcf02d8d3/Dockerfile#L27-L30), however Rockcraft doesn't support templating or build arguments
* the working directory is not set
  * Python imports from /app will fail
  * the pod sets the command explicitly, thus bypassing Pebble. Rocraft doesn't allow setting env variables or the working directory in this case.

As a workaround, we'll use a Dockerfile wrapper to set image fields that Rockcraft would not allow otherwise.

Related PR: https://github.com/canonical/k8s-workflows/pull/39

Fixes: https://github.com/canonical/k8s-snap/issues/1278

Follow-up tasks:

* drop the `--args rawfile` argument that was introduced to accommodate Pebble: https://github.com/canonical/k8s-snap/blob/5eb2f5a059d432e7f970cb799d7be67915cd4869/src/k8s/pkg/k8sd/features/localpv/localpv.go#L38
* update the helm chart from k8s-snap
* since this ensures that the rock image can be used as a drop-in replacement of the upstream image, we'll probably no longer have to maintain this fork
* propose the task.yaml fix upstream

petrutlucian94 committed:
Use the right image entry point

Rockcraft doesn't allow setting the image entry point, forcing
Pebble instead.

For this reason, the rock images cannot be used as drop-in
replacements: https://github.com/canonical/k8s-snap/issues/1278#issuecomment-2791412856

As a workaround, we'll use a Dockerfile wrapper to set image
fields that Rockcraft would not allow otherwise.

Related PR: https://github.com/canonical/k8s-workflows/pull/39

petrutlucian94 committed:
donotmerge: use wip workflow

petrutlucian94 committed:
minor fixes.

petrutlucian94 committed:
Try the disk cleanup action on arm64

We're running out of disk space while copying the image.

petrutlucian94 committed:
Use self-hosted arm runner

petrutlucian94 commented:
Need to fix the arm job, it's running out of disk space.

petrutlucian94 committed:
Start a tmate action on failure

petrutlucian94 committed:
Revert "Use self-hosted arm runner"

This reverts commit be9611d97a525f0c15f5918b25550a06ce5a1f66.

petrutlucian94 committed:
Revert "Try the disk cleanup action on arm64"

This reverts commit 5375597d792424999e61a0e7b0f42781aca8f63e.

petrutlucian94 committed:
Fix Python venv failure

A recent commit was merged directly (without a PR) and broke the
rock build:

https://github.com/canonical/rawfile-localpv/commit/c02f1608ae180588bf9cbca97799260d9a9a3a62

The symink interferes with virtualenv:

```
:: + python3 -m venv /root/parts/rawfile-deps/install
:: Error: Unable to create directory '/root/parts/rawfile-deps/install/bin'
Failed to run the build script for part 'rawfile-deps'.
Launching shell on build environment...
rockcraft-rawfile-localpv-arm64-259507 ../project# ls -lh /root/parts/rawfile-deps/install/bin
lrwxrwxrwx 1 root root 7 Feb  5 16:08 /root/parts/rawfile-deps/install/bin -> usr/bin
```

We'll restore the old workflow to unblock the image build.

petrutlucian94 committed:
Disable tmate

petrutlucian94 commented on .github/workflows/pull_request.yaml:
```diff
@@ -7,16 +7,41 @@ on:
   build-and-push-arch-specifics:
     name: Build Rocks and Push Arch Specific Images
-    uses: canonical/k8s-workflows/.github/workflows/build_rocks.yaml@main
+    uses: canonical/k8s-workflows/.github/workflows/build_rocks.yaml@lpetrut/b64_dockerfile
```
TODO: use the main branch after the following PR merges: https://github.com/canonical/k8s-workflows/pull/39

petrutlucian94 committed:
disable rock cache

Commit:
Adjust noble container build to properly test pebble (#27)

* adjust arm64 ld-linx

* test using rockcraft version

* Test the start error experiences a missing kube config

* create symlink after rawfile-deps are built

* Symlink /bin/sh in prime step

* Rollback to jammy where things seems to work better

addyess commented:
I'm not 100% sold on this approach of replacing the pebble entrypoint. Remind me again why this is necessary?  Are there places where the pebble entrypoint is causing an issue (such as the resizer not working) and we can't replace the entry point by adjusting the Pod spec instead?

petrutlucian94 commented:
> I'm not 100% sold on this approach of replacing the pebble entrypoint. Remind me again why this is necessary? Are there places where the pebble entrypoint is causing an issue (such as the resizer not working) and we can't replace the entry point by adjusting the Pod spec instead?

To adjust the [pod spec](https://github.com/canonical/rawfile-localpv/blob/main/templates/task.yaml) we'd have to make changes in rawfile-localpv and the consensus was that we're trying to avoid maintaining project forks. Rock images should be drop-in replacements of the upstream images.

addyess commented:
i agree we'd HAVE to change the [pod-spec](https://github.com/canonical/rawfile-localpv/blob/main/templates/task.yaml#L31-L33) to be something like:

```yaml
      command:
        - exec
        - /bin/sh
        - -c
```

right?  Yes, this would alter the `rawfile-localpv` to make this possible.  The same is done in other places in this repo to alter the rawfile container.  I just don't know if we should jump to ignoring pebble.  I think that's a too far when we've guaranteed these images are created by rockcraft.

petrutlucian94 commented:
We already spent a lot of time trying to find workarounds to accommodate Pebble, a component that brings absolutely no value when there's just one process per container (99% of the time). Pebble just interferes with the way in which Kubernetes launches and manages container processes, being entirely redundant.

We agreed that we no longer want to fork projects or the helm charts, but instead use this workaround until Rockcraft makes Pebble optional.

petrutlucian94 commented:
TODO:

* ~~define working directory~~
* ~~pass image tag and repo~~
* ~~avoid hardcoded data directory in task yaml~~

petrutlucian94 committed:
set work dir

petrutlucian94 commented:
There's another problem that needs to be addressed: the pod template ignores the data directory from the helm chart, but instead uses this hardcoded path: https://github.com/canonical/rawfile-localpv/blob/3bcdc388a39d9352c1cbe652965c268bcf02d8d3/templates/task.yaml#L14

Meanwhile, we're setting the following (` /var/snap/k8s/common/rawfile-storage`): https://github.com/canonical/k8s-snap/blob/ebb0996c48e1106d2b0d01428f9ed6c816c0ee27/src/k8s/pkg/k8sd/features/localpv/localpv.go#L49-L51

For this reason, the resize will fail: 

```
$ sudo k8s kubectl logs -n kube-system $(sudo k8s kubectl get pods -n kube-system | grep task| awk '{print $1}')
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
  File "<string>", line 7, in expand_rawfile
  File "/app/rawfile_util.py", line 33, in img_file
    return Path(metadata(volume_id)["img_file"])
KeyError: 'img_file'
```

petrutlucian94 committed:
Update task.yaml, using configured host path data directory

The helm chart allows using a configurable data directory host path,
however task.yaml uses hard-coded value.

For this reason, the resize fails when using non-default paths
such as the one used by k8s-snap: /var/snap/k8s/common/rawfile-storage

This change will use the configured path instead of the hard-coded
one.

petrutlucian94 commented:
I'm closing this for now, feel free to reopen at a later time if needed. The decision was to wait for a change in Rockcraft.
